062-support-tiered-storage.md

```diff
@@ -0,0 +1,117 @@
+```
+
+### RemoteStorageManager (RSM)
+
+The RSM config only supports custom type. The config allows users to specify the RSM class information.
+
+Because the `RemoteStorageManager` dependency is provided by Kafka plugin, the implementation could vary case by case. It’s important to allow user to pass in additional configuration values within the supplied prefix range. The custom configuration can be supplied via  `spec.kafka.config` with the predefined prefix.
+
+### RemoteLogMetadataManager (RLMM)
+
+For simplicity, the proposal doesn't add customization for RemoteLogMetadataManager. If tiered storage is enabled, the default implementation using `TopicBasedRemoteLogMetadataManager` will be used, and corresponding configuration will be appended. 
+
+Since the default RLMM will create an internal Kafka client to talk to the server, there might be some additional configuration needed, like the SSL related setting. Those config can be put in `.spec.kafka.config` directly using the `rlmm.config.` prefix. Example can be found below.
+
+### Example configuration
+
+The below config define a sample configuration for tiered storage setup, using a `custom` RSM type.
+```
+kafka:
+  tieredStorage:
+    type: custom
+    rsm.className: com.example.kafka.tiered.storage.s3.S3RemoteStorageManager
+    rsm.classPath:  /opt/kafka/plugins/tiered-storage-s3/*
+    rsm.impl.prefix: rsm.config.
+
+  config:
+    ...
+    rlmm.config.remote.log.metadata.common.client.bootstrap.servers: broker-bootstrap.com:9091
+    rlmm.config.remote.log.metadata.common.client.security.protocol: SSL
+```
+
+The configuration above will get translated to the below Kafka broker config:
+```
+remote.log.storage.system.enable: true
+remote.log.storage.manager.class.name: com.example.kafka.tiered.storage.s3.S3RemoteStorageManager
+remote.log.storage.manager.class.path: /opt/kafka/plugins/tiered-storage-s3/*
+remote.log.storage.manager.impl.prefix: rsm.config.
+# The RLMM config will be added by default
+remote.log.metadata.manager.impl.prefix: rlmm.config.
+remote.log.metadata.manager.class.name: org.apache.kafka.server.log.remote.metadata.storage.TopicBasedRemoteLogMetadataManager
+# The RLMM config below are passed down from kafka config spec.
+rlmm.config.remote.log.metadata.common.client.bootstrap.servers: broker-bootstrap.com:9091
+rlmm.config.remote.log.metadata.common.client.security.protocol: SSL
+```
+
+### Validation
+
+There are certain [limitations](https://cwiki.apache.org/confluence/display/KAFKA/Kafka+Tiered+Storage+Early+Access+Release+Notes) for tiered storage enabled cluster.
+
+With Strimzi operator, we can validate the configuration before apply the change. Here are some validations we can apply:
+
+* Tiered storage should not be disabled at cluster level when there is any existing topic enabled tiered storage.
+
+* Tiered storage should not be enabled on a cluster with JBOD.
+
+* Tiered storage should not be applied for compact topic.
```

Comment:
I think that validating all of these would be quite complicated - the JBOD storage should be quite simple. But the rest would not. You should either take the pragmatic way and leave it to the user or you will need to figure out and elaborate on how to do such validation.

```suggestion
There are certain [limitations](https://cwiki.apache.org/confluence/display/KAFKA/Kafka+Tiered+Storage+Early+Access+Release+Notes) for tiered storage enabled cluster.
For example:
* Tiered storage should not be disabled at cluster level when there is any existing topic enabled tiered storage.
* Tiered storage should not be enabled on a cluster with JBOD.
* Tiered storage should not be applied for compact topic.

Some of those - such as the JBOD storage support - are easy to validate in the operator and will be validated automatically.
Others - such as not enabling it on compacted topics or not disabling it while any topics using tiered storage still exist would add too much complexity and it will be the responsibility of the users to take care of them.
```

Reply:
As a user of strimzi I'd like to see the validation of these items within strimzi tbh. At least for the limitations that concern topic-level items there should be validation if these are controlled by the UTO. I can see a case for leaving the validation of non-managed topics to the user on the other hand. 

This "split" of the validation would cover our use-case well as we have a policy to have any "stateful" (not really stateful but somewhat statey) topics defined as `KafkaTopics` while allowing other topics to be directly generated if it doesn't matter at all if they are lost.

Reply:
Doing these validations:
* Would likely be a big setback for the tiered storage support in Strimzi in terms of this proposal being approved and implemented. Some of them might not even be possible to check.
* A lot of these are limitations of the Kafka implementation and I guess many of them are expected to be implemented in the future. So implementing them would be also possibly a waste of resources.

> This "split" of the validation would cover our use-case well as we have a policy to have any "stateful" (not really stateful but somewhat statey) topics defined as KafkaTopics while allowing other topics to be directly generated if it doesn't matter at all if they are lost.

This might apply in your case, but I do not think this can be generalized. Many important topics are created outside of the Topic Operator - Connect topics, Kafka Streams API topics etc.

Reply:
> * A lot of these are limitations of the Kafka implementation and I guess many of them are expected to be implemented in the future. So implementing them would be also possibly a waste of resources.

It may be worth checking out the KIPs to find out which limitations are intended to remain and which are just transient until tiered storage is feature complete. I'll see if I can find anything that I would consider to be worth it to validate in strimzi. 

> This might apply in your case, but I do not think this can be generalized.

I agree that the "statefulness" discussion may be specific to our use-case, I do believe that the expectation of other users is similar however. IMO if you are not managing a topic with the UTO then you aren't expecting strimzi to handle anything about it. In general however I think the impact of TS on the TO hasn't been handled, however this is mentioned in a separate review thread so we can probably see what happens there.

Reply:
Thinking about this a bit more:
* The proposal now links to the list of limitations in Kafka Wiki. But as these might get edited with future releases, it would be good to copy the limitations into this proposal to understand the state as it was when this proposal was written.
* I asked and all of the limitations (apart from the client compatibility) are expected to be fixed before it is production-ready in Kafka. This is worth mentioning here as well.
* As for the Topic Operator ... I did not tried it, but I expect Kafka to handle these limitations and validate them when managing topics? So any special treatment in Topic Operator will be just changing the Kafka exception for a Strimzi exception and not sure it really adds value.

If we are not sure users will understand all the limitations, we can also consider adding a feature gate to protect the Tiered Storage:
* If it is disabled, the `tieredStorage` section from the Kafka CR will be essentially ignored
* By enabling it, users confirm they understand the limitations
* Once it moves out of early access in Kafka and the limitations are fixed, we can enable the feature gate by default.

I'm not saying this should be added to the proposal right away -> just adding it as an option for the consideration by others @strimzi/maintainers, @Staniel and @siegenthalerroger